Fixes #26991 - Add Ansible Collections

sjha4 opened the pull request:
To test, copy the change in katello.yaml.example to foreman/config/settings.plugins.d/katello.yaml and restart server after installing the binding gem. The one-VM setup for Pulp3 has the pulp_ansible enabled.

- Read documentation for the pulp-ansible plugin here : https://pulp-ansible.readthedocs.io/en/latest/
- Read API documentation for the plugin here: https://pulp-ansible.readthedocs.io/en/latest/restapi.html
- Ruby binding gem for pulp_ansible can be found here: https://rubygems.org/gems/pulp_ansible_client

jlsherrill commented on app/lib/actions/katello/repository/metadata_generate.rb:
```diff
@@ -15,14 +15,16 @@ def plan(repository, options = {})
                         :dependency => dependency,
-                        :repository_creation => repository_creation)
+                        :repository_creation => repository_creation,
+                        :publication_content_type => publication_content_type)
```
any reason this needs to be determined here?  can it just be handled in the pulp3 orchestration layer?

jlsherrill commented on app/models/katello/repository.rb:
```diff
@@ -145,15 +150,15 @@ class Repository < Katello::Model
              :ostree_upstream_sync_policy, :ostree_upstream_sync_depth, :deb_releases, :deb_components, :deb_architectures,
              :ignore_global_proxy, :ssl_ca_cert_id, :ssl_ca_cert, :ssl_client_cert, :ssl_client_cert_id, :ssl_client_key_id,
-             :ssl_client_key, :ignorable_content, :description, :docker_tags_whitelist, :to => :root
+             :ssl_client_key, :ignorable_content, :description, :docker_tags_whitelist, :ansible_collection_whitelist, :to => :root
```
lets not delegate new attributes to root, and just call them on root directly?

sjha4 replied:
Sticking with this for now based on chats with Justin.. Will refactor if required later.

jlsherrill commented on app/services/katello/pulp3/ansible_collection.rb:
```diff
@@ -0,0 +1,47 @@
+            response = response.as_json.with_indifferent_access
+            yielder.yield response[:results]
+            page_opts[:page] += 1
+          end
```
can we DRY this up between file and ansible_collection?

jlsherrill commented on app/services/katello/pulp3/repository/ansible_collection.rb:
```diff
@@ -0,0 +1,95 @@
+              create_remote
+            end
+          end
+        end
```
I think we could DRY this up as well

jlsherrill commented on katello.gemspec:
```diff
@@ -44,6 +44,7 @@ Gem::Specification.new do |gem|
   gem.add_dependency "anemone"
   gem.add_dependency "pulpcore_client"
   gem.add_dependency "pulp_file_client", "< 1.0"
+  gem.add_dependency "pulp_ansible_client", "0.2.0b1.dev0.1560866833"
```
Reminder for me to do a packaging PR before this gets merged

jlsherrill replied:
PR opened: https://github.com/theforeman/foreman-packaging/pull/3858

jlsherrill commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -482,6 +484,7 @@ def construct_repo_from_params(repo_params)
       root.upstream_username = repo_params[:upstream_username] if repo_params.key?(:upstream_username)
       root.upstream_password = repo_params[:upstream_password] if repo_params.key?(:upstream_password)
       root.ignorable_content = repo_params[:ignorable_content] if root.yum? && repo_params.key?(:ignorable_content)
+      root.ansible_collection_whitelist = repo_params[:ansible_collection_whitelist] if root.ansible_collection?
```
As part of adding this new type, it'd be great to identify places where we can make the code more extensible, and hardcode type-specific entries less often.  Maybe lets get together and brainstorm and document some ways of doing it, even if we do it at a later time.

sjha4 replied:
In this case specifically, we could just ignore the check and let the field be set to nil?

sjha4 commented:
[test katello]

sjha4 commented:
[test katello]

jlsherrill commented on app/lib/actions/pulp3/orchestration/repository/generate_metadata.rb:
```diff
@@ -5,11 +5,12 @@ module Repository
         class GenerateMetadata < Pulp3::Abstract
           def plan(repository, smart_proxy, options = {})
             options[:contents_changed] = (options && options.key?(:contents_changed)) ? options[:contents_changed] : true
+            publication_content_type = !(repository.content_type == "ansible_collection")
```
could we add some smarts into the repository type registration to know about this?  Maybe something like:

```
Katello::RepositoryTypeManager.register(::Katello::Repository::FILE_TYPE) do
  allow_creation_by_user true
  service_class Katello::Pulp::Repository::File
  pulp3_service_class Katello::Pulp3::Repository::File
  pulp3_plugin 'pulp_file'
  pulp3_uses_publications

  content_type Katello::FileUnit, :pulp2_service_class => ::Katello::Pulp::FileUnit, :pulp3_service_class => ::Katello::Pulp3::FileUnit, :user_removable => true
end

```

to indicate that file type uses publications (and anything that doesn't declare that, does not)

sean797 commented on app/controllers/katello/api/v2/repositories_controller.rb:
```diff
@@ -54,6 +54,7 @@ def custom_index_relation(collection)
       param :deb_architectures, String, :desc => N_("comma separated list of architectures to be synched from deb-archive")
       param :ignore_global_proxy, :bool, :desc => N_("if true, will ignore the globally configured proxy when syncing")
       param :ignorable_content, Array, :desc => N_("List of content units to ignore while syncing a yum repository. Must be subset of %s") % RootRepository::IGNORABLE_CONTENT_UNIT_TYPES.join(",")
+      param :ansible_collection_whitelist, String, :desc => N_("Name of collection to sync from URL")
```
Calling this `ansible_collection_whitelist` implies it should be an array to me.

sean797 replied:
Also, the description should say `Ansible Collection` IMHO

sean797 replied:
Aplogies, one more thing - Are you going to have a `whitelist` function for other content types with Pulp 3? If so, it would make sence to use a more generic name.

jlsherrill replied:
we do also have a whitelist for docker too, thats not a bad idea!  We'd have to depreciate the old docker one, and move it over to the common one.  Although then if pulp ever supported whitelisting of multiple types within a repository, that wouldn't work too well.... hrmmmm  Let me think about it

jlsherrill replied:
so i chatted with samir about this, and i initially liked the idea, but if we ever have some plugin that have multiple whitelists for content types, or some other kind of format, we'd need other parameters for that.  

I think we're going to have this explosion in repository attributes regardless that we need to work on dealing with that.  As we continue to add content types it may become apparent that this was a bad decision, and we can revisit it :)

sean797 commented on db/migrate/20190617142328_create_katello_ansible_collections.rb:
```diff
@@ -0,0 +1,26 @@
+    end
+
+    add_index :katello_ansible_collections, :pulp_id, :unique => true, :name => 'katello_ansible_collections_pulp_id_index'
+    add_index :katello_ansible_collections, [:id, :pulp_id, :name, :version, :namespace], :name => 'katello_ansible_collections_fields_index'
```
Do we really need to index every attribute?

sjha4 replied:
Removed indexing for all fields other than pulp_id.

sean797 commented on app/models/katello/ansible_collection.rb:
```diff
@@ -0,0 +1,32 @@
+
+    scoped_search :on => :name, :complete_value => true
+    scoped_search :on => :namespace, :complete_value => true
+    scoped_search :on => :checksum
```
Can you add a search for `version` ?

sjha4 commented:
[test katello]

sjha4 commented:
@jlsherrill : updated pulp_ansible_client gem to 0.2.0b1.dev0.1561471371

sjha4 commented:
Pinning gem to <=0.2.0b1.dev0.1561386170

jlsherrill commented on db/migrate/20190618034438_add_ansible_collection_whitelist_to_katello_root_repositories.rb:
```diff
@@ -0,0 +1,5 @@
+class AddAnsibleCollectionWhitelistToKatelloRootRepositories < ActiveRecord::Migration[5.2]
+  def change
+    add_column :katello_root_repositories, :ansible_collection_whitelist, :string
```
I think we want this to be :text as to not have a 255 char limit

jlsherrill commented:
![image](https://user-images.githubusercontent.com/395077/60363393-638da880-99b1-11e9-87e5-38ea86d4dc4c.png)

* Could we 'suggest' the default galaxy url? Kinda like what we do with docker.
* white list seems required (i can create without a whitelist but can't sync), we should probably indicate that if that is intentional in pulp
* an example of what to type into the whitelist would be very helpful, when i try to do multiple moduels i can't seem to get it to work and the errors i've seen imply i can paste 'requirements' in 
* should whitelist be a larger text box so you can put more text in it easily?

sjha4 commented:
> Could we 'suggest' the default galaxy url? Kinda like what we do with docker.

Added a help text `"URL of the ansible content sharing hub you want to sync. Example: https://galaxy.ansible.com"`
> white list seems required (i can create without a whitelist but can't sync), we should probably indicate that if that is intentional in pulp

Added a model validation.
> an example of what to type into the whitelist would be very helpful, when i try to do multiple moduels i can't seem to get it to work and the errors i've seen imply i can paste 'requirements' in

Added a help text ` Name / Comma-separated names of ansible collection(s) you want to sync from the galaxy. ex: newswangerd.collection_demo, devoperate.base`
> should whitelist be a larger text box so you can put more text in it easily?

Made the field a text-area field.

jlsherrill commented:
Looking good!  

* When syncing i end up with the message 'New Content Units: 0'  even when new units had been synced.
* I'm getting an error when re-syncing a repo, the task starts but fails with error: Missing the required parameter 'repository_version_href' when calling RepositoriesApi.repositories_versions_read

sjha4 commented:
@jlsherrill: I have a few issues open with pulp to address these:
1. https://pulp.plan.io/issues/5023 : Ansible Collection sync does not have data in the progress report
2. https://pulp.plan.io/issues/4997 : Pulp3 Ansible Collection remote doesn't seem to be able to sync more than 1 repo
3. https://pulp.plan.io/issues/5018 : Add artifact sha details to the content ansible collection list API

These should take care of the issues with sync and indexing.

sjha4 commented:
> I'm getting an error when re-syncing a repo, the task starts but fails with error: Missing the required parameter 'repository_version_href' when calling RepositoriesApi.repositories_versions_read . This may not be ansible specific

This is ansible specific. There is a fix for this available but that is post the binding changes so it's not working with this PR. The issue tracking this was https://pulp.plan.io/issues/4920 wherein new repo versions were not created on re-sync. That is what is causing the error you see on re-sync.

sjha4 commented:
Thanks @jlsherrill , @sean797 ..
